src/coreclr/jit/optimizer.cpp

```diff
@@ -9056,9 +9100,15 @@ GenTree* OptBoolsDsc::optIsBoolComp(OptTestInfo* pOptTest)
     assert(pOptTest->testTree->gtOper == GT_JTRUE || pOptTest->testTree->gtOper == GT_RETURN);
     GenTree* cond = pOptTest->testTree->AsOp()->gtOp1;
 
-    // The condition must be "!= 0" or "== 0"
+    // The condition must be "!= 0" or "== 0" or >=0 or <0
+
+    if (!cond->OperIs(GT_EQ, GT_NE, GT_LT, GT_GE))
+    {
+        return nullptr;
+    }
 
-    if ((cond->gtOper != GT_EQ) && (cond->gtOper != GT_NE))
+    // we don't optimize this statements because we might delete them (e.g. array range checks)
+    if (cond->OperIs(GT_LT, GT_GE) && (m_b1->bbFlags != BBF_IMPORTED || m_b2->bbFlags != BBF_IMPORTED))
```

Comment:
This check looks a little odd to me. We usually don't care about `BBF_IMPORTED` later on in the jit. I wonder if we should be checking something else here instead to convey the meaning better.

Do you recall what lead you to add this?

Reply:
array range checks instructions (indexer < 0, indexer < array.Length) are added to the user code, but later they get deleted during the Range Check Elimination step. When those instructions are passed through the boolean optimization step they get optimized with user code, so they're no longer deleted. That's why I added the BBF_IMPORTED filter so I optimize only user code, I know the filter is not constraining enough, but I didn't find some flag/property that could filter only array range check instructions.

Reply:
This is to fix a regression and not to fix some correctness issue? If so then it's likely a similar effect to what I mention below with jump threading. An upstream opt inhibits a downstream one.

If my experiment to move `OptOptimizeBools` to the end of the optimization pass pans out, then this won't be necessary.

Should have some data on it soon.

Reply:
Yes it's to fix regression (not to fix correctness issue)